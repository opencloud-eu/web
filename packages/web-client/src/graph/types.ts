import type { AxiosInstance, ResponseType } from 'axios'
import type { Configuration } from './generated'

export interface GraphFactoryOptions {
  axiosClient: AxiosInstance
  config: Configuration
}

export interface GraphRequestOptions {
  headers?: Record<string, string>
  params?: Record<string, string>
  signal?: AbortSignal
  responseType?: ResponseType
}
