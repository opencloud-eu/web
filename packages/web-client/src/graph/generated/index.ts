/* tslint:disable */
/* eslint-disable */
/**
 * Libre Graph API
 * Libre Graph is a free API for cloud collaboration inspired by the MS Graph API.
 *
 * The version of the OpenAPI document: v1.0.7
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

