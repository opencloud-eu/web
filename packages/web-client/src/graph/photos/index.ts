export * from './photo'
export * from './types'
