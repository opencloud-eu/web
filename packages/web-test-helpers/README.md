# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
