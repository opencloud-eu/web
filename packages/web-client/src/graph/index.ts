import { AxiosInstance } from 'axios'
import { Configuration } from './generated'
import { type GraphUsers, UsersFactory } from './users'
import { type GraphGroups, GroupsFactory } from './groups'
import { ApplicationsFactory, GraphApplications } from './applications'
import { DrivesFactory, GraphDrives } from './drives'
import { DriveItemsFactory, GraphDriveItems } from './driveItems'
import { TagsFactory, GraphTags } from './tags'
import { ActivitiesFactory, GraphActivities } from './activities'
import { PermissionsFactory, GraphPermissions } from './permissions'
import { PhotoFactory, GraphPhotos } from './photos'

export interface Graph {
  activities: GraphActivities
  applications: GraphApplications
  tags: GraphTags
  drives: GraphDrives
  driveItems: GraphDriveItems
  users: GraphUsers
  groups: GraphGroups
  permissions: GraphPermissions
  photos: GraphPhotos
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  return <Graph>{
    activities: ActivitiesFactory({ axiosClient, config }),
    applications: ApplicationsFactory({ axiosClient, config }),
    tags: TagsFactory({ axiosClient, config }),
    drives: DrivesFactory({ axiosClient, config }),
    driveItems: DriveItemsFactory({ axiosClient, config }),
    users: UsersFactory({ axiosClient, config }),
    groups: GroupsFactory({ axiosClient, config }),
    permissions: PermissionsFactory({ axiosClient, config }),
    photos: PhotoFactory({ axiosClient, config })
  }
}
