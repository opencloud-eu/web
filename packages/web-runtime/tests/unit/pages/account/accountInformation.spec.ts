import AccountInformation from '../../../../src/pages/account/accountInformation.vue'
import { defaultComponentMocks, defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import {
  Extension,
  ExtensionPoint,
  OptionsConfig,
  useExtensionRegistry
} from '@opencloud-eu/web-pkg'
import { User } from '@opencloud-eu/web-client/graph/generated'

const $route = {
  meta: {
    title: 'Some Title'
  }
}

const selectors = {
  editUrlButton: '[data-testid="account-page-edit-url-btn"]',
  logoutButton: '[data-testid="account-page-logout-url-btn"]',
  accountPageInfo: '.account-page-info',
  groupNames: '[data-testid="group-names"]',
  groupNamesEmpty: '[data-testid="group-names-empty"]'
}

describe('account information page', () => {
  describe('public link context', () => {
    it('should render a limited view', async () => {
      const { wrapper } = getWrapper({ isUserContext: false, isPublicLinkContext: true })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('header section', () => {
    describe('edit url button', () => {
      it('should be displayed if defined via config', () => {
        const { wrapper } = getWrapper({
          accountEditLink: { href: '/' }
        })

        const editUrlButton = wrapper.find(selectors.editUrlButton)
        expect(editUrlButton.html()).toMatchSnapshot()
      })
      it('should not be displayed if not defined via config', () => {
        const { wrapper } = getWrapper()

        const editUrlButton = wrapper.find(selectors.editUrlButton)
        expect(editUrlButton.exists()).toBeFalsy()
      })
    })
  })

  describe('account information section', () => {
    it('displays basic user information', () => {
      const { wrapper } = getWrapper({
        user: mock<User>({
          onPremisesSamAccountName: 'some-username',
          displayName: 'some-displayname',
          mail: 'some-email',
          memberOf: []
        })
      })

      const accountPageInfo = wrapper.find(selectors.accountPageInfo)
      expect(accountPageInfo.html()).toMatchSnapshot()
    })

    describe('group membership', () => {
      it('displays message if not member of any groups', () => {
        const { wrapper } = getWrapper()

        const groupNamesEmpty = wrapper.find(selectors.groupNamesEmpty)
        expect(groupNamesEmpty.exists()).toBeTruthy()
      })
      it('displays group names', () => {
        const { wrapper } = getWrapper({
          user: mock<User>({
            memberOf: [{ displayName: 'one' }, { displayName: 'two' }, { displayName: 'three' }]
          })
        })

        const groupNames = wrapper.find(selectors.groupNames)
        expect(groupNames.html()).toMatchSnapshot()
      })
    })

    describe('Logout from all devices link', () => {
      it('should render the logout from active devices if logoutUrl is provided', async () => {
        const { wrapper } = getWrapper()

        expect(wrapper.find('[data-testid="logout"]').exists()).toBe(true)
      })
      it("shouldn't render the logout from active devices if logoutUrl isn't provided", async () => {
        const { wrapper } = getWrapper()

        ;(wrapper.vm as any).logoutUrl = undefined
        expect(wrapper.find('[data-testid="logout"]').exists()).toBe(true)
      })
      it('should use url from configuration manager', async () => {
        const { wrapper } = getWrapper()

        const logoutButton = wrapper.find(selectors.logoutButton)
        expect(logoutButton.attributes('href')).toBe('https://account-manager/logout')
      })
    })
  })
})

function getWrapper({
  user = mock<User>({ memberOf: [] }),
  accountEditLink = undefined,
  isPublicLinkContext = false,
  isUserContext = true,
  extensionPoints = [],
  extensions = []
}: {
  user?: User
  accountEditLink?: OptionsConfig['accountEditLink']
  isPublicLinkContext?: boolean
  isUserContext?: boolean
  extensionPoints?: ExtensionPoint<Extension>[]
  extensions?: Extension[]
} = {}) {
  const plugins = defaultPlugins({
    piniaOptions: {
      userState: { user },
      authState: {
        userContextReady: isUserContext,
        publicLinkContextReady: isPublicLinkContext
      },
      configState: {
        options: {
          logoutUrl: 'https://account-manager/logout',
          ...(accountEditLink && { accountEditLink })
        }
      }
    }
  })

  const { getExtensionPoints, requestExtensions } = useExtensionRegistry()
  vi.mocked(getExtensionPoints).mockReturnValue(extensionPoints)
  vi.mocked(requestExtensions).mockReturnValue(extensions)

  const mocks = {
    ...defaultComponentMocks(),
    $route
  }

  mocks.$clientService.graphAuthenticated.users.getMe.mockResolvedValue(mock<User>({ id: '1' }))

  return {
    mocks,
    wrapper: mount(AccountInformation, {
      global: {
        plugins,
        mocks,
        provide: mocks
      }
    })
  }
}
