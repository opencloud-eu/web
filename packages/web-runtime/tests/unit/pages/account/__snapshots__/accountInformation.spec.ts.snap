// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account information page > account information section > displays basic user information 1`] = `
"<div class="account-table account-page-info mt-6">
  <table data-v-381278b9="" class="oc-table-simple">
    <colgroup>
      <col class="w-auto md:w-[30%]">
      <col class="w-auto md:w-[40%]">
      <col class="w-auto md:w-[30%]">
    </colgroup>
    <thead class="oc-thead border-b sr-only">
      <tr data-v-98c4aba8="">
        <th data-v-27e13406="" data-v-c53469a5="" class="oc-table-cell text-left align-middle oc-th aria-[sort]:cursor-pointer">Information name</th>
        <th data-v-27e13406="" data-v-c53469a5="" class="oc-table-cell text-left align-middle oc-th aria-[sort]:cursor-pointer">Information value</th>
      </tr>
    </thead>
    <tbody>
      <tr data-v-98c4aba8="">
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">
          <div>Profile picture</div>
          <div class="text-sm text-role-on-surface-variant">Max. 10MB, JPG, PNG</div>
        </td>
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">
          <div class="[&amp;_.cropper-crop-box]:!outline-1 [&amp;_.cropper-crop-box]:!outline-role-outline [&amp;_.cropper-view-box]:!rounded-[50%] [&amp;_.cropper-crop-box]:!rounded-[50%] [&amp;_.cropper-line]:!bg-role-outline [&amp;_.cropper-point]:!bg-role-outline mb-2"><input class="invisible avatar-file-input" type="file" accept="image/jpeg, image/png">
            <div class="flex flex-col items-center"><span class="vue-avatar--wrapper oc-avatar flex justify-center items-center shrink-0 text-center rounded-[50%] select-none mb-4" style="width: 128px; height: 128px; line-height: 128px; background-color: #9C27B0; font-size: 51px; font-family: Helvetica, Arial, sans-serif; color: white;" width="128" aria-hidden="true" focusable="false" data-test-user-name="some-displayname"><span class="avatar-initials" style="color: white;">SD</span></span>
              <div>
                <div class="oc-button-group"><button type="button" class="oc-button-secondary oc-button-outline oc-button-secondary-outline gap-2 justify-center text-sm min-h-3 oc-button cursor-pointer disabled:opacity-60 disabled:cursor-default">
                    <!--v-if-->
                    <!-- @slot Content of the button --> Upload
                  </button>
                  <!--v-if-->
                </div>
              </div>
            </div>
            <!--v-if-->
            <!--v-if-->
          </div>
        </td>
      </tr>
      <tr data-v-98c4aba8="" class="account-page-info-username">
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Username</td>
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">some-username</td>
      </tr>
      <tr data-v-98c4aba8="" class="account-page-info-displayname">
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">First and last name</td>
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">some-displayname</td>
      </tr>
      <tr data-v-98c4aba8="" class="account-page-info-email">
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Email</td>
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">some-email</td>
      </tr>
      <!--v-if-->
      <tr data-v-98c4aba8="" class="account-page-info-groups">
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Group memberships</td>
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td" data-testid="group-names"><span data-testid="group-names-empty">You are not part of any group</span></td>
      </tr>
      <tr data-v-98c4aba8="" class="account-page-logout-all-devices">
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Logout from active devices</td>
        <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td" data-testid="logout"><a href="https://account-manager/logout" target="_blank" class="oc-button-secondary oc-button-raw oc-button-secondary-raw gap-2 justify-center text-base min-h-4 no-hover oc-button cursor-pointer disabled:opacity-60 disabled:cursor-default" data-testid="account-page-logout-url-btn">
            <!--v-if-->
            <!-- @slot Content of the button --> <span>Show devices</span>
          </a></td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`account information page > account information section > group membership > displays group names 1`] = `"<td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td" data-testid="group-names"><span>one, two, three</span></td>"`;

exports[`account information page > header section > edit url button > should be displayed if defined via config 1`] = `
"<a href="/" target="_blank" class="oc-button-secondary oc-button-outline oc-button-secondary-outline gap-2 justify-center text-base min-h-4 oc-button cursor-pointer disabled:opacity-60 disabled:cursor-default" data-testid="account-page-edit-url-btn">
  <!--v-if-->
  <!-- @slot Content of the button --> <span class="oc-icon box-content inline-block align-baseline [&amp;_svg]:block size-5.5"><svg data-testid="inline-svg-stub" src="icons/edit-fill.svg" transform-source="(svg) => {
      if (__props.accessibleLabel !== &quot;&quot;) {
        const title = document.createElement(&quot;title&quot;);
        title.setAttribute(&quot;id&quot;, svgTitleId.value);
        title.appendChild(document.createTextNode(__props.accessibleLabel));
        svg.insertBefore(title, svg.firstChild);
      };
      return svg;
    }" aria-hidden="true" focusable="false" class="size-5.5"></svg></span> <span>Edit</span>
</a>"
`;

exports[`account information page > public link context > should render a limited view 1`] = `
"<div id="account-information">
  <div class="flex justify-between items-center w-full">
    <h1 class="mt-0">Account Information</h1>
    <!--v-if-->
  </div>
  <div class="account-table account-page-info mt-6">
    <table data-v-381278b9="" class="oc-table-simple">
      <colgroup>
        <col class="w-auto md:w-[30%]">
        <col class="w-auto md:w-[40%]">
        <col class="w-auto md:w-[30%]">
      </colgroup>
      <thead class="oc-thead border-b sr-only">
        <tr data-v-98c4aba8="">
          <th data-v-27e13406="" data-v-c53469a5="" class="oc-table-cell text-left align-middle oc-th aria-[sort]:cursor-pointer">Information name</th>
          <th data-v-27e13406="" data-v-c53469a5="" class="oc-table-cell text-left align-middle oc-th aria-[sort]:cursor-pointer">Information value</th>
        </tr>
      </thead>
      <tbody>
        <tr data-v-98c4aba8="">
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">
            <div>Profile picture</div>
            <div class="text-sm text-role-on-surface-variant">Max. 10MB, JPG, PNG</div>
          </td>
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">
            <div class="[&amp;_.cropper-crop-box]:!outline-1 [&amp;_.cropper-crop-box]:!outline-role-outline [&amp;_.cropper-view-box]:!rounded-[50%] [&amp;_.cropper-crop-box]:!rounded-[50%] [&amp;_.cropper-line]:!bg-role-outline [&amp;_.cropper-point]:!bg-role-outline mb-2"><input class="invisible avatar-file-input" type="file" accept="image/jpeg, image/png">
              <div class="flex flex-col items-center"><span class="vue-avatar--wrapper oc-avatar flex justify-center items-center shrink-0 text-center rounded-[50%] select-none mb-4" style="width: 128px; height: 128px; line-height: 128px; background-color: #b82015; font-size: 51px; font-family: Helvetica, Arial, sans-serif; color: white;" width="128" aria-hidden="true" focusable="false" data-test-user-name=""><span class="avatar-initials" style="color: white;"></span></span>
                <div>
                  <div class="oc-button-group"><button type="button" class="oc-button-secondary oc-button-outline oc-button-secondary-outline gap-2 justify-center text-sm min-h-3 oc-button cursor-pointer disabled:opacity-60 disabled:cursor-default">
                      <!--v-if-->
                      <!-- @slot Content of the button --> Upload
                    </button>
                    <!--v-if-->
                  </div>
                </div>
              </div>
              <!--v-if-->
              <!--v-if-->
            </div>
          </td>
        </tr>
        <tr data-v-98c4aba8="" class="account-page-info-username">
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Username</td>
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td"></td>
        </tr>
        <tr data-v-98c4aba8="" class="account-page-info-displayname">
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">First and last name</td>
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td"></td>
        </tr>
        <tr data-v-98c4aba8="" class="account-page-info-email">
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Email</td>
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td"><span>No email has been set up</span></td>
        </tr>
        <!--v-if-->
        <tr data-v-98c4aba8="" class="account-page-info-groups">
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td">Group memberships</td>
          <td data-v-27e13406="" class="oc-table-cell text-left align-middle oc-td" data-testid="group-names"><span data-testid="group-names-empty">You are not part of any group</span></td>
        </tr>
        <!--v-if-->
      </tbody>
    </table>
  </div>
</div>"
`;
